relay-profiling/src/sample.rs

```diff
@@ -95,6 +123,107 @@ impl Profile {
+                Some(stack) => !stack.is_empty(),
+                None => true,
+            };
+
+            if !is_active {
+                continue;
+            }
+
+            if let Some(range) = active_ranges.get_mut(&sample.thread_id) {
+                range.end = i + 1;
+            } else {
+                active_ranges.insert(sample.thread_id, i..i + 1);
+            }
+        }
+
+        self.samples = self
+            .samples
+            .drain(..)
+            .enumerate()
+            .filter(|(i, sample)| {
+                if let Some(range) = active_ranges.get(&sample.thread_id) {
+                    range.contains(i)
+                } else {
+                    false
+                }
+            })
+            .map(|(_, sample)| sample)
+            .collect();
+    }
+
+    /// Removes a sample when it's the only sample on its thread
+    fn remove_single_samples_per_thread(&mut self) {
+        let mut sample_count_by_thread_id: HashMap<u64, u32> = HashMap::new();
+
+        for sample in &self.samples {
+            *sample_count_by_thread_id
+                .entry(sample.thread_id)
+                .or_default() += 1;
+        }
+
+        // Only keep data from threads with more than 1 sample so we can calculate a duration
+        sample_count_by_thread_id.retain(|_, count| *count > 1);
+
+        self.samples
+            .retain(|sample| sample_count_by_thread_id.contains_key(&sample.thread_id));
+    }
+
+    fn check_samples(&self) -> bool {
+        for sample in &self.samples {
+            if self.stacks.get(sample.stack_id).is_none() {
+                return false;
+            }
+        }
+        true
+    }
+
+    fn check_stacks(&self) -> bool {
+        for stack in &self.stacks {
+            for frame_id in stack {
+                if self.frames.get(*frame_id).is_none() {
+                    return false;
+                }
+            }
+        }
+        true
+    }
+
+    fn is_above_max_duration(&self) -> bool {
+        if let Some(sample) = &self.samples.last() {
+            return sample.elapsed_since_start_ns > MAX_PROFILE_DURATION_NS;
+        }
+        false
+    }
+
+    fn cleanup_thread_metadata(&mut self) {
+        if let Some(thread_metadata) = &mut self.thread_metadata {
+            let mut thread_ids: HashSet<String> = HashSet::new();
+            for sample in &self.samples {
+                thread_ids.insert(sample.thread_id.to_string());
+            }
+            thread_metadata.retain(|thread_id, _| thread_ids.contains(thread_id));
+        }
+    }
+
+    fn cleanup_queue_metadata(&mut self) {
+        if let Some(queue_metadata) = &mut self.queue_metadata {
+            let mut queue_addresses: HashSet<&String> = HashSet::new();
+            for sample in &self.samples {
+                if let Some(queue_address) = &sample.queue_address {
+                    queue_addresses.insert(queue_address);
+                }
+            }
```

Comment:
Also here,

```suggestion
            let queue_addresses = self
                .samples
                .iter()
                .filter_map(|sample| sample.queue_address.as_ref())
                .collect::<HashSet<_>>();

```
What do you think?

Reply:
I like that thanks.